plugin/storage/es/spanstore/reader.go

```diff
@@ -217,6 +255,173 @@ func (s *SpanReader) bucketToStringArray(buckets []*elastic.AggregationBucketKey
+	//           { "range":  { "duration": { "gte": traceQuery.DurationMin, "lte": traceQuery.DurationMax }}},
+	//           { "nested" : {
+	//             "path" : "tags",
+	//             "query" : {
+	//                 "bool" : {
+	//                     "must" : [
+	//                     { "match" : {"tags.key" : "traceQuery.Tags.key"} },
+	//                     { "match" : {"tags.value" : "traceQuery.Tags.value"} }
+	//                     ]
+	//                 }}}}
+	//         ]
+	//       }
+	//     },
+	//     "aggs": { "traceIDs" : { "terms" : {"size": numOfTraces,"field": "traceID" }}}
+	// }
+	aggregation := s.buildTraceIDAggregation(traceQuery.NumTraces)
+	boolQuery := s.buildFindTraceIDsQuery(traceQuery)
+
+	jaegerIndices := s.findIndices(traceQuery)
+
+	searchService := s.client.Search(jaegerIndices...).
+		Type(spanType).
+		Size(0). // set to 0 because we don't want actual documents.
+		Aggregation(traceIDAggregation, aggregation).
+		Query(boolQuery)
+
+	searchResult, err := searchService.Do(s.ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "Search service failed")
+	}
+
+	bucket, found := searchResult.Aggregations.Terms(traceIDAggregation)
+	if !found {
+		err = errors.New("Could not find aggregation of traceIDs")
+		return nil, err
+	}
+
+	traceIDBuckets := bucket.Buckets
+	return s.bucketToStringArray(traceIDBuckets)
+}
+
+func (s *SpanReader) buildTraceIDAggregation(numOfTraces int) elastic.Aggregation {
+	return elastic.NewTermsAggregation().
+		Size(numOfTraces).
+		Field(traceIDField)
+}
+
+func (s *SpanReader) buildFindTraceIDsQuery(traceQuery *spanstore.TraceQueryParameters) elastic.Query {
+	boolQuery := elastic.NewBoolQuery()
+
+	//add duration query
+	if traceQuery.DurationMax != 0 || traceQuery.DurationMin != 0 {
+		durationQuery := s.buildDurationQuery(traceQuery.DurationMin, traceQuery.DurationMax)
+		boolQuery.Must(durationQuery)
+	}
+
+	//add startTime query
+	startTimeQuery := s.buildStartTimeQuery(traceQuery.StartTimeMin, traceQuery.StartTimeMax)
+	boolQuery.Must(startTimeQuery)
+
+	//add process.serviceName query
+	if traceQuery.ServiceName != "" {
+		serviceNameQuery := s.buildServiceNameQuery(traceQuery.ServiceName)
+		boolQuery.Must(serviceNameQuery)
+	}
+
+	//add operationName query
+	if traceQuery.OperationName != "" {
+		operationNameQuery := s.buildOperationNameQuery(traceQuery.OperationName)
+		boolQuery.Must(operationNameQuery)
+	}
+
+	//add tags query (must be nested) TODO: add log tags query
+	for k, v := range traceQuery.Tags {
+		tagQuery := s.buildTagQuery(k, v)
+		boolQuery.Must(tagQuery)
+	}
+	return boolQuery
+}
+
+func (s *SpanReader) buildDurationQuery(durationMin time.Duration, durationMax time.Duration) elastic.Query {
+	minDurationMicros := durationMin.Nanoseconds() / int64(time.Microsecond/time.Nanosecond)
+	maxDurationMicros := (time.Hour * 24).Nanoseconds() / int64(time.Microsecond/time.Nanosecond)
```

Comment:
hmm what's the point of this? wouldn't I have to set it again in this fn and this fn would be the only time i use it?